src/libutil/ustring.cpp

```diff
@@ -108,38 +149,64 @@ template<unsigned BASE_CAPACITY, unsigned POOL_SIZE> struct TableRepMap {
 #endif
         size_t pos = hash & mask, dist = 0;
         for (;;) {
-            if (entries[pos] == 0)
+            ustring::TableRep* e = entries[pos];
+            if (e == 0)
                 return 0;
-            if (entries[pos]->hashed == hash)
-                return entries[pos]->c_str();
+            if (e->hashed == hash)
+                return e->c_str();
             ++dist;
             pos = (pos + dist) & mask;  // quadratic probing
         }
     }
 
     const char* insert(string_view str, uint64_t hash)
     {
+        OIIO_ASSERT((hash & ustring::duplicate_bit) == 0);  // can't happen?
         ustring_write_lock_t lock(mutex);
         size_t pos = hash & mask, dist = 0;
+        bool duplicate_hash = false;
         for (;;) {
-            if (entries[pos] == 0)
+            ustring::TableRep* e = entries[pos];
+            if (e == 0)
                 break;  // found insert pos
-            if (entries[pos]->hashed == hash
-                && entries[pos]->length == str.length()
-                && !strncmp(entries[pos]->c_str(), str.data(), str.length())) {
-                // same string is already inserted, return the one that is
-                // already in the table
-                return entries[pos]->c_str();
+            if (e->hashed == hash) {
+                duplicate_hash = true;
+                if (e->length == str.length()
+                    && !strncmp(e->c_str(), str.data(), str.length())) {
+                    // same string is already inserted, return the one that is
+                    // already in the table
+                    return e->c_str();
+                }
             }
             ++dist;
             pos = (pos + dist) & mask;  // quadratic probing
         }
 
         ustring::TableRep* rep = make_rep(str, hash);
-        entries[pos]           = rep;
+
+        // If we encountered another ustring with the same hash (if one
+        // exists, it would have hashed to the same address so we would have
+        // seen it), set the duplicate bit in the rep's hashed field.
+        if (duplicate_hash) {
```

Comment:
If you change the polarity of the indicator bit, you can pass "duplicate_hash" into make_rep, and then it can just store the pointer from the pool alloc in the "hashed" member, and all you have to do here is inc the collisions counter (or the inc and dbug print could be moved into make_rep as well. -- if you do that, you need to do another pobe for a valid "pos" here based on the pointer bits rather than the original hash bits.